### DISCUSSION TITLE
Schema Validation while also sending 405 status code

### DISCUSSION THREAD

**Discussion by hrishikesh-k** [fastify/fastify#4719] [Q&A] [answered]
Related issue: https://github.com/fastify/fastify/issues/917

For (good) reasons, Fastify is suppressing HTTP 405. I was working around that by doing something like:

```js
fastify.route({
  config: {
    allowedMethods: ['POST']
  },
  method: [
    'COPY',
    'DELETE',
    'GET',
    'HEAD',
    'LOCK',
    'MKCOL',
    'MOVE',
    'OPTIONS',
    'PATCH',
    'POST',
    'PROPFIND',
    'PROPPATCH',
    'PUT',
    'SEARCH',
    'TRACE',
    'UNLOCK'
  ],
  onRequest: (onRequestRequest, _onRequestReply, onRequestDone) => {
    if (onRequestRequest.routeConfig.allowedMethods.some(allowedMethod => {
      return allowedMethod === onRequestRequest.method
    })) {
      onRequestDone()
    } else {
      // ApiError is an extension of FastifyError
      onRequestDone(new ApiError(`${onRequestRequest.method} method is not allowed on ${onRequestRequest.url}`, 405))
    }
  },
  schema: {
    body: {}
  }
})
```

Now, when I add some schema for request validation to this, even if I send a POST request,  Fastify fails with `FST_ERR_ROUTE_BODY_VALIDATION_SCHEMA_NOT_SUPPORTED`, the description being: `Body validation schema for GET:<url> route is not supported!`.

Is there a way where I can keep the schema validation while being able to send a 405 status at the same time? In worst case, I'd have to drop the 405 code as schema is more important, but I'd rather want to keep both if possible.

**Comment by mcollina** [comment]
you should use a preHandler hook instead.

**Reply by hrishikesh-k** [reply]
Thanks. Do you mean using `preHandler` instead of `onRequest` for the 405 handling? If so, I'm not sure how that would allow the schema to be supported, unless I'm missing something.

Do you mind elaborating a bit? 🙂

**Reply by mcollina** [reply]
I don’t fully understand what you are trying to achieve here… but onRequest hooks are executedb before validation.

**Reply by hrishikesh-k** [reply]
That is exactly what I need. I need to send a 405 even before the validation happens (as I don't want to spend time validating if the method is not allowed). However, inspite of having that code in the `onRequest` hook, Fastify complains that `Body validation schema for GET:<url> route is not supported!`. For example consider this situation:

- Someone makes a non-POST request to my endpoint.
- I wish to show them a 405 status code, hopefully without any validation as that's not required.
- However, someone makes a POST request to that endpoint.
- I wish to validate the request body and process the request normally.

So my thinking was like, allow the handler to accept all methods, and add an extra property for just the allowed methods. When the request starts, check if the handler's allowed methods has the method used by the request. If yes, proceed with validation and other processing. If not, send a 405 and close the request.

Does that make sense?

**Reply by mcollina** [reply, marked as answer]
This workflow is not currently supported by fastify, you need different route definitions.

**Reply by hrishikesh-k** [reply]
I kinda thought that would be the case. Thanks for confirming!
